active/0051-packaging.md

```diff
@@ -0,0 +1,338 @@
++ dead modules elimination inside of the `addons/` (application would be the
+  main entry point and unused modules are removed only from `addons/`)
++ dead modules elimination inside of the `app/`
+  + removing unused components and helpers (requires analysing templates)
+  + removing unused initializers/services (this likely entails work on
+    dependency injection layer as we would need access to a resolver resolution
+    map)
++ tree-shaking (Rollup-like tree-shaking where we include *only* the code that is
+  used)
+
+`Linker` would be able to take an `exclude` list of modules as a parameter.
+Although, valuable in some situations, it should be clearly marked as advanced
+API. It should be used as a last resort and serve as an "escape hatch".
+
+It would make sense to implement `Linker` as a strategy. Developers would be
+able to "opt in"/"opt out" of optimising behaviour.
+
+## Deprecating `app.import` API
+
+Ember applications which choose to use `Linker` strategy should be able to
+remove usages of `app.import`.
+
+## Tools
+
+With growing complexity of Ember applications, it is crucial to provide more
+insights into final assets.
+
+Main goals are:
+
+- report raw/uglified/compressed asset sizes;
+  [broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser)
+- find source code duplication across your javascript assets (enables you to
+  fine tune code splitting parameters to reduce bundle invalidation rates as
+  well as improve repeat page load performance)
+
+# How We Teach This
+
+This is a backward compatible change to the existing Ember CLI ecosystem. In
+order to teach users how to use `package` API, we need to update the API docs
+with a section for this and the best practices of when to use this. A more
+general purpose blog post could be beneficial as well.
+
+# Drawbacks
+
+There are several potential drawbacks that are worth noting.
+
+_Build performance_. There is minimal overhead in instantiating strategies and
+calling methods on them and I believe this approach shouldn't degrade build
+performance.
+
+_A note on add-ons_. Add-ons don't rely on the way Ember CLI does bundling. That
+means existing build system continues to work as expected and add-ons won't have
+to change their implementation.
+
+# Alternatives
+
+This RFC allows us to customise packaging when needed.
+[Webpack](https://webpack.js.org) has become very popular in solving this
+similar problem. One could implement a `package` function that would use Webpack
+for packaging. Ultimately, we need something that is aware of how Ember apps are
+assembled and how Ember apps utilise dependency injection that takes advantage
+of existing tools. The long term plan is to have a dependency graph that is
+aware of application structure so can avoid the "wall of configuration" that
+other asset packaging systems are susceptible to.
+
+# Unresolved questions
+
++ Will it increase build time?
++ Should we introduce the same API on add-on level?
```

Comment:
Generally, yes. I was toying around with the idea of maybe exposing some of the dep graph APIs on the add-on level. Having said that, it might be too far in the future